File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        # return one for root because root is always a good node
        return 1 + self.rec(root.left, root.val) + self.rec(root.right, root.val)
    
    def rec (self, root: TreeNode, maxVal) -> int:
        if root == None:
            return 0
        
        if root.val >= maxVal:
            if root.val > maxVal:
                maxVal = root.val
            return 1 + self.rec(root.left, maxVal) + self.rec(root.right, maxVal)
        
        return self.rec(root.left, maxVal) + self.rec(root.right, maxVal)
        

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var countGoodNodes = function(root, max) {
    if (!root) return 0
    
    const thisCount = root.val >= max ? 1 : 0
    max = Math.max(root.val, max)
    
    const leftCount = countGoodNodes(root.left, max)
    const rightCount = countGoodNodes(root.right, max)
    
    return leftCount + rightCount + thisCount
}

var goodNodes = function(root) {
    return countGoodNodes(root, Number.NEGATIVE_INFINITY)
}